File: backend/app/models/user_settings.py
"""
用户设置模型
"""
from datetime import datetime, timezone

from sqlmodel import Field, SQLModel

from app.core.config import settings


class UserSettings(SQLModel, table=True):
    """用户设置"""
    __tablename__ = "user_settings"

    id: int = Field(default=1, primary_key=True)  # 单例设置，固定ID为1
    rsshub_url: str = Field(default="https://rsshub.app", description="RSSHub基础URL")
    fetch_interval_minutes: int = Field(default=720, description="RSS获取间隔（分钟），默认12小时")
    auto_refresh: bool = Field(default=True, description="是否自动刷新")
    show_description: bool = Field(default=True, description="是否显示RSS描述")
    items_per_page: int = Field(default=50, description="每页显示条目数")
    show_entry_summary: bool = Field(default=True, description="时间线是否显示摘要")
    open_original_mode: str = Field(default="system", description="打开原文方式: 'system' | 'window'")

    # 时间过滤相关设置
    enable_date_filter: bool = Field(default=True, description="是否启用时间过滤")
    default_date_range: str = Field(default="30d", description="默认时间范围: '1d', '2d', '3d', '7d', '30d', '90d', '180d', '365d', 'all'")
    time_field: str = Field(default="inserted_at", description="时间字段: 'published_at' 或 'inserted_at'")
    max_auto_title_translations: int = Field(default=10, description="自动标题翻译的并发/批量上限")
    mark_as_read_range: str = Field(default="current", description="一键已读范围: 'current', '3d', '7d', '30d', 'all'")

    # AI config and features
    summary_api_key: str = Field(default="", description="AI summary API key")
    summary_base_url: str = Field(default=settings.glm_base_url, description="AI summary base URL")
    summary_model_name: str = Field(default=settings.glm_model, description="AI summary model")
    translation_api_key: str = Field(default="", description="AI translation API key")
    translation_base_url: str = Field(default=settings.glm_base_url, description="AI translation base URL")
    translation_model_name: str = Field(default=settings.glm_model, description="AI translation model")
    ai_auto_summary: bool = Field(default=False, description="Auto-generate summaries")
    ai_auto_title_translation: bool = Field(default=False, description="Auto-translate titles")
    ai_title_display_mode: str = Field(default="original-first", description="Title display mode")
    ai_translation_language: str = Field(default="zh", description="Title translation language")

    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Config:
        """SQLModel配置"""
        table_name = "user_settings"

File: backend-node/src/services/userSettings.ts
/**
 * User settings service
 */

import { getDatabase } from '../db/session.js';
import { UserSettings } from '../db/models.js';

const FETCH_INTERVAL_MIN = 5;
const FETCH_INTERVAL_MAX = 1440;
const DEFAULT_FETCH_INTERVAL_MINUTES = 720;

export class InvalidUserSettingsUpdateError extends Error {
  invalidKeys: string[];

  constructor(invalidKeys: string[]) {
    super(`Invalid user settings fields: ${invalidKeys.join(', ')}`);
    this.name = 'InvalidUserSettingsUpdateError';
    this.invalidKeys = invalidKeys;
  }
}

type TableInfoRow = { name: string };

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function isSafeSqlIdentifier(value: string): boolean {
  // Conservative: SQLite identifiers for columns in this project are snake_case.
  return /^[a-z_][a-z0-9_]*$/.test(value);
}

/**
 * Normalize fetch interval to valid range
 */
export function normalizeFetchInterval(value: number | null | undefined): number {
  if (value === null || value === undefined) {
    return DEFAULT_FETCH_INTERVAL_MINUTES;
  }
  if (value < FETCH_INTERVAL_MIN) {
    return FETCH_INTERVAL_MIN;
  }
  if (value > FETCH_INTERVAL_MAX) {
    return FETCH_INTERVAL_MAX;
  }
  return value;
}

export class UserSettingsService {
  private db = getDatabase();
  private updatableColumns: Set<string> | null = null;

  private getUpdatableColumns(): Set<string> {
    if (this.updatableColumns) {
      return this.updatableColumns;
    }
    const tableInfo = this.db.pragma('table_info(user_settings)') as TableInfoRow[];
    const columns = new Set(tableInfo.map((row) => row.name));
    columns.delete('id');
    columns.delete('created_at');
    columns.delete('updated_at');
    this.updatableColumns = columns;
    return columns;
  }

  /**
   * Get user settings, create default if not exists
   */
  getSettings(): UserSettings {
    let settings = this.db.prepare('SELECT * FROM user_settings WHERE id = 1').get() as UserSettings | undefined;

    if (!settings) {
      // Create default settings
      const now = new Date().toISOString();
      this.db.prepare(`
        INSERT INTO user_settings (
          id, rsshub_url, fetch_interval_minutes, auto_refresh, show_description,
          items_per_page, show_entry_summary, open_original_mode, enable_date_filter,
          default_date_range, time_field, max_auto_title_translations, mark_as_read_range,
          details_panel_mode, summary_api_key, summary_base_url, summary_model_name,
          translation_api_key, translation_base_url, translation_model_name,
          ai_auto_summary, ai_auto_title_translation, ai_title_display_mode,
          ai_translation_language, summary_prompt_preference, translation_prompt_preference,
          outbound_proxy_mode, outbound_proxy_url, created_at, updated_at
        ) VALUES (
          1, 'https://rsshub.app', 720, 1, 1, 50, 1, 'system', 1, '30d', 'inserted_at',
          10, 'current', 'docked', '', 'https://open.bigmodel.cn/api/paas/v4/', 'glm-4-flash',
          '', 'https://open.bigmodel.cn/api/paas/v4/', 'glm-4-flash', 0, 0, 'original-first',
          'zh', '', '', 'system', '', ?, ?
        )
      `).run(now, now);

      settings = this.db.prepare('SELECT * FROM user_settings WHERE id = 1').get() as UserSettings;
    }

    // Normalize fetch interval
    const normalizedInterval = normalizeFetchInterval(settings.fetch_interval_minutes);
    if (settings.fetch_interval_minutes !== normalizedInterval) {
      this.db.prepare('UPDATE user_settings SET fetch_interval_minutes = ?, updated_at = ? WHERE id = 1')
        .run(normalizedInterval, new Date().toISOString());
      settings.fetch_interval_minutes = normalizedInterval;
    }

    return settings;
  }

  /**
   * Update user settings
   */
  updateSettings(updates: unknown): UserSettings {
    const settings = this.getSettings();
    const now = new Date().toISOString();

    if (!isPlainObject(updates)) {
      throw new Error('Invalid settings payload: expected an object');
    }

    // Build update query dynamically
    const allowedColumns = this.getUpdatableColumns();
    const updateFields: string[] = [];
    const values: Array<string | number | boolean | null> = [];
    const invalidKeys: string[] = [];

    for (const [key, value] of Object.entries(updates)) {
      if (value === undefined) continue;
      if (key === 'id' || key === 'created_at' || key === 'updated_at') {
        invalidKeys.push(key);
        continue;
      }
      if (!isSafeSqlIdentifier(key) || !allowedColumns.has(key)) {
        invalidKeys.push(key);
        continue;
      }
      if (
        value !== null &&
        typeof value !== 'string' &&
        typeof value !== 'number' &&
        typeof value !== 'boolean'
      ) {
        invalidKeys.push(key);
        continue;
      }

      updateFields.push(`${key} = ?`);
      values.push(value);
    }

    if (invalidKeys.length > 0) {
      throw new InvalidUserSettingsUpdateError(invalidKeys);
    }

    if (updateFields.length > 0) {
      updateFields.push('updated_at = ?');
      values.push(now);
      values.push(1); // WHERE id = 1

      const query = `UPDATE user_settings SET ${updateFields.join(', ')} WHERE id = ?`;
      this.db.prepare(query).run(...values);
    }

    return this.getSettings();
  }

  /**
   * Get configured RSSHub URL
   */
  getRSSHubUrl(): string {
    const settings = this.getSettings();
    return settings.rsshub_url.replace(/\/$/, ''); // Remove trailing slash
  }

  /**
   * Update RSSHub URL
   */
  updateRSSHubUrl(url: string): UserSettings {
    // Ensure URL format is correct
    if (!url.startsWith('http://') && !url.startsWith('https://')) {
      url = 'https://' + url;
    }

    return this.updateSettings({ rsshub_url: url });
  }
}

// Global instance
export const userSettingsService = new UserSettingsService();
